### PR TITLE
Never release store using CancellableThreads

### LINKED ISSUE #45237: org.elasticsearch.cluster.SimpleDataNodesIT » testAutoExpandReplicasAdjustedWhenDataNodeJoins
## Example build failure
https://scans.gradle.com/s/exu7mrkb4htbs/tests/kyv2y2z3r4v7m-cdlwdmcdh66q6

## Reproduction line 
does not reproduce locally
```
./gradlew :server:integTest --tests "org.elasticsearch.cluster.SimpleDataNodesIT.testAutoExpandReplicasAdjustedWhenDataNodeJoins" -Dtests.seed=A4B7524EEBD18A4 -Dtests.security.manager=true -Dtests.locale=en-FK -Dtests.timezone=Asia/Atyrau -Dcompiler.java=12 -Druntime.java=11
```

## Example relevant log:
```
org.apache.lucene.store.AlreadyClosedException: FileLock invalidated by an external force: NativeFSLock(path=/var/lib/jenkins/workspace/elastic+elasticsearch+master+multijob+fast+part1/server/build/testrun/integTest/temp/org.elasticsearch.cluster.SimpleDataNodesIT_A4B7524EEBD18A4-001/tempDir-002/node_t1/node.lock,impl=sun.nio.ch.FileLockImpl[0:9223372036854775807 exclusive invalid],creationTime=2019-08-05T10:23:13.542911Z)Close stacktrace
at __randomizedtesting.SeedInfo.seed([A4B7524EEBD18A4:4CD2DA2EA74514BC]:0)
at org.apache.lucene.store.NativeFSLockFactory$NativeFSLock.ensureValid(NativeFSLockFactory.java:178)
at org.elasticsearch.env.NodeEnvironment.assertEnvIsLocked(NodeEnvironment.java:1122)
at org.elasticsearch.env.NodeEnvironment.nodeDataPaths(NodeEnvironment.java:880)
at org.elasticsearch.test.InternalTestCluster$NodeAndClient.markNodeDataDirsAsPendingForWipe(InternalTestCluster.java:959)
at org.elasticsearch.test.InternalTestCluster$NodeAndClient.close(InternalTestCluster.java:943)
at org.elasticsearch.core.internal.io.IOUtils.close(IOUtils.java:104)
at org.elasticsearch.core.internal.io.IOUtils.close(IOUtils.java:86)
at org.elasticsearch.test.InternalTestCluster.close(InternalTestCluster.java:801)
at org.elasticsearch.test.ESIntegTestCase.afterInternal(ESIntegTestCase.java:545)
at org.elasticsearch.test.ESIntegTestCase.cleanUpCluster(ESIntegTestCase.java:1962)
at jdk.internal.reflect.GeneratedMethodAccessor9.invoke(Unknown Source)
at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.lang.reflect.Method.invoke(Method.java:566)
at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1750)
at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
at org.apache.lucene.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:49)
at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
at org.apache.lucene.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:48)
at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:817)
at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:468)
at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:947)
at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:832)
at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:883)
at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:894)
at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
at org.apache.lucene.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:41)
at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:54)
at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
at java.lang.Thread.run(Thread.java:834)
```

### PR DESCRIPTION
**Pull request by dnhatn**
Today we can release a Store using CancellableThreads. If we are holding the last reference, then we will verify the [node lock ](https://github.com/elastic/elasticsearch/blob/42e8b36f5e2d134d0212baa5932e5a99f8c00386/server/src/main/java/org/elasticsearch/env/NodeEnvironment.java#L940)before deleting the store. Checking node lock performs some I/O on [FileChannel](https://github.com/apache/lucene-solr/blob/2eb493d1700d59845ac120dcc485556b7e7fb422/lucene/core/src/java/org/apache/lucene/store/NativeFSLockFactory.java#L182). If the current thread is interrupted, then the channel will be [closed](https://github.com/openjdk-mirror/jdk7u-jdk/blob/master/src/share/classes/java/nio/channels/spi/AbstractInterruptibleChannel.java#L165) and the node lock will also be invalid.

Previously, we only run isolated tasks under the primary shard permit using CancellableThreads. However, this is no longer true after #45136 where we can trigger other actions if we fail to sync retention leases.

Closes #45237

### CODE REVIEW

**Comment by dnhatn** [comment]
We are sending requests using CancellableThreads for example https://github.com/elastic/elasticsearch/blob/30339acab21734d228c08fd384fe3746b1c34ecd/server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java#L485-L487. If we are unable to submit a request, then that failure would execute other actions using CancellableThreads. I don't think we need to send requests using  CancellableThreads. We used it before because we send and wait for the response synchronously.

@ywelsch @henningandersen WDYT?

File: server/src/test/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java [hunk 1]
```diff
@@ -1486,4 +1487,15 @@ public void testPeerRecoveryTrimsLocalTranslog() throws Exception {
...
+    public void testCancelRecoveryWithAutoExpandReplicas() {
+        internalCluster().startMasterOnlyNode();
+        assertAcked(client().admin().indices().prepareCreate("test")
+            .setSettings(Settings.builder().put(IndexMetaData.SETTING_AUTO_EXPAND_REPLICAS, "0-all"))
+            .setWaitForActiveShards(ActiveShardCount.NONE));
+        internalCluster().startNode();
+        internalCluster().startNode();
+        client().admin().cluster().prepareReroute().setRetryFailed(true).get();
+        assertAcked(client().admin().indices().prepareDelete("test")); // cancel recoveries
+    }
```
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java]: Can we check that the recoveries are no longer running afterwards?

**Review by henningandersen — COMMENTED** [review]
Thanks @dnhatn. I left a few comments to consider/comment on.

**Comment by dnhatn** [comment]
> I am not really sure about this, but it seems the culprit is here and maybe we could just spawn the decRef on the generic thread pool and wait for it to complete?

@henningandersen Yes, this should work and it's a safe choice for now. I've updated this PR to adopt this suggestion. We can make other changes in the follow-up PRs. Can you please take another look? Thank you.

**Review by henningandersen — APPROVED** [review]
LGTM.

Thanks @dnhatn and we should definitely pick up the other changes in follow-ups to make recovery async.

**Comment by dnhatn** [comment]
Thanks @henningandersen.

**Pull request merged**